File: src/com/medium/BinaryTreeLevelOrderTraversal_102.java
/**
 * 
 */
package com.medium;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class BinaryTreeLevelOrderTraversal_102 {
    // 2019年8月20日
    // Use BFS
    // Time Complexity: O(n)
    // Space Complexity:O(n)
    // Runtime: 1 ms, faster than 89.54%
    // Memory Usage: 36.2 MB, less than 100.00%
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> mResult = new ArrayList<>();

        if (root == null) {
            return mResult;
        }

        Queue<TreeNode> mQueue = new LinkedList<TreeNode>();
        mQueue.add(root);

        while (!mQueue.isEmpty()) {

            List<Integer> mList = new ArrayList<>();
            int size = mQueue.size();

            for (int i = 0; i < size; i++) {
                TreeNode mNode = mQueue.poll();
                if (mNode != null) {
                    mList.add(mNode.val);
                }
                if (mNode.left != null) {
                    mQueue.add(mNode.left);
                }

                if (mNode.right != null) {
                    mQueue.add(mNode.right);
                }
            }

            mResult.add(mList);

        }

        return mResult;
    }

    // 2019年8月20日
    // Use DFS (pre-order)
    // Time Complexity: O(n)
    // Space Complexity:O(n)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 36.5 MB, less than 100.00%
    public List<List<Integer>> levelOrder2(TreeNode root) {
        List<List<Integer>> mResult = new ArrayList<>();

        DFS(root, 0, mResult);

        return mResult;
    }

    private void DFS(TreeNode root, int mDepth, List<List<Integer>> mResult) {
        if (root == null) {
            return;
        }
        if (mDepth >= mResult.size()) {
            mResult.add(new LinkedList<Integer>());
        }
        mResult.get(mDepth).add(root.val);
        DFS(root.left, mDepth + 1, mResult);
        DFS(root.right, mDepth + 1, mResult);
    }
}

File: Golang_src/medium/binaryTreeLevelOrderTraversal_102.go
package medium

// Use BFS:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2.8 MB, less than 100.00%
func levelOrder(root *TreeNode) [][]int {
	result := [][]int{}
	if root == nil {
		return result
	}

	queue := []*TreeNode{root}

	for len(queue) > 0 {
		size := len(queue)
		tmp := []int{}
		for i := 0; i < size; i++ {
			node := queue[0]
			queue = queue[1:]
			if node != nil {
				tmp = append(tmp, node.Val)
			}
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}
		result = append(result, tmp)
	}

	return result
}

// Use DFS:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 3 MB, less than 34.47%
func levelOrder2(root *TreeNode) [][]int {
	result := [][]int{}
	if root == nil {
		return result
	}

	var dfs func(node *TreeNode, level int)
	dfs = func(node *TreeNode, level int) {
		if node == nil {
			return
		}
		if level == len(result) {
			result = append(result, []int{})
		}
		result[level] = append(result[level], node.Val)
		dfs(node.Left, level+1)
		dfs(node.Right, level+1)
	}

	dfs(root, 0)

	return result
}
